File: eeg-backend/neuro_yogic/complexity.py
"""
complexity.py
=============
Non-linear complexity / entropy features for EEG epochs — the reference oracle
for the C# port in ``src/NeuroYogic.SignalProcessing/Dsp/Complexity.cs``.

Motivation
----------
The deepest contemplative states (Ekagra / Niruddha) are distinguished by
*non-oscillatory* dynamics that band-power and phase measures capture poorly.
These four measures index that structure directly:

  lziv          : Lempel-Ziv complexity (LZ76, normalised) — regularity of the
                  median-binarised signal.
  higuchi_fd    : Higuchi fractal dimension — self-similarity / roughness.
  sample_entropy: Sample entropy (m=2, r=0.2·std) — unpredictability.
  perm_entropy  : Permutation entropy (order=3) — ordinal-pattern diversity.

Each is computed per channel then averaged across channels, matching the
band-power convention in feature_extractor.py. Implementations are deliberately
hand-rolled in pure numpy (no antropy dependency) so the algorithm is
transparent and portable bit-for-behaviour to C#.
"""
from math import log, log2

import numpy as np


def lempel_ziv(x: np.ndarray) -> float:
    """Normalised Lempel-Ziv complexity (LZ76) of the median-binarised signal.

    Returns c(n) · log2(n) / n, where c(n) is the LZ76 substring count.
    """
    n = len(x)
    if n < 2:
        return 0.0
    med = float(np.median(x))
    s = [1 if v > med else 0 for v in x]

    # LZ76 substring-counting (Kaspar & Schuster).
    i, c, l, k, k_max = 0, 1, 1, 1, 1
    while True:
        if s[i + k - 1] == s[l + k - 1]:
            k += 1
            if l + k > n:
                c += 1
                break
        else:
            if k > k_max:
                k_max = k
            i += 1
            if i == l:
                c += 1
                l += k_max
                if l + 1 > n:
                    break
                i = 0
                k = 1
                k_max = 1
            else:
                k = 1
    return float(c * log2(n) / n)


def higuchi_fd(x: np.ndarray, kmax: int = 10) -> float:
    """Higuchi fractal dimension via the log-log slope of curve length L(k)."""
    n = len(x)
    if n < 4:
        return 0.0
    kmax = min(kmax, n // 2)
    lnk = []
    lnl = []
    for k in range(1, kmax + 1):
        lm = []
        for m in range(k):
            n_m = (n - m - 1) // k
            if n_m <= 0:
                continue
            acc = 0.0
            for i in range(1, n_m + 1):
                acc += abs(x[m + i * k] - x[m + (i - 1) * k])
            norm = (n - 1) / (n_m * k)
            lm.append((acc * norm) / k)
        if lm:
            lnl.append(log(float(np.mean(lm))))
            lnk.append(log(1.0 / k))
    if len(lnk) < 2:
        return 0.0
    # Least-squares slope of lnl vs lnk.
    return float(_slope(lnk, lnl))


def sample_entropy(x: np.ndarray, m: int = 2, r_coef: float = 0.2) -> float:
    """Sample entropy SampEn(m, r) with r = r_coef · std(x) (population std)."""
    n = len(x)
    if n < m + 2:
        return 0.0
    r = r_coef * float(np.std(x))
    if r <= 0.0:
        return 0.0

    def count(mm: int) -> int:
        # N - m templates for both m and m+1 (aligned convention).
        count_n = n - m
        c = 0
        for i in range(count_n):
            for j in range(i + 1, count_n):
                # Chebyshev distance over the mm-length template.
                d = 0.0
                for t in range(mm):
                    diff = abs(x[i + t] - x[j + t])
                    if diff > d:
                        d = diff
                if d <= r:
                    c += 1
        return c

    b = count(m)
    a = count(m + 1)
    if a == 0 or b == 0:
        return 0.0
    return float(-log(a / b))


def perm_entropy(x: np.ndarray, order: int = 3, delay: int = 1) -> float:
    """Normalised permutation entropy of ordinal patterns (Bandt-Pompe)."""
    n = len(x)
    span = delay * (order - 1)
    if n - span <= 1:
        return 0.0
    counts: dict = {}
    for i in range(n - span):
        window = [x[i + t * delay] for t in range(order)]
        pattern = tuple(_argsort_stable(window))
        counts[pattern] = counts.get(pattern, 0) + 1
    total = sum(counts.values())
    pe = 0.0
    for c in counts.values():
        p = c / total
        pe -= p * log2(p)
    denom = log2(_factorial(order))
    return float(pe / denom) if denom > 0 else 0.0


def complexity_features(filtered: np.ndarray) -> dict:
    """Per-channel complexity measures averaged across channels.

    Parameters
    ----------
    filtered : ndarray (n_channels, n_samples) — the band-pass+notch signal.

    Returns
    -------
    dict with keys: lziv, higuchi_fd, sample_entropy, perm_entropy.
    """
    filtered = np.atleast_2d(filtered)
    lz, hf, se, pe = [], [], [], []
    for ch in filtered:
        ch = np.asarray(ch, dtype=np.float64)
        lz.append(lempel_ziv(ch))
        hf.append(higuchi_fd(ch))
        se.append(sample_entropy(ch))
        pe.append(perm_entropy(ch))
    return {
        "lziv": float(np.mean(lz)),
        "higuchi_fd": float(np.mean(hf)),
        "sample_entropy": float(np.mean(se)),
        "perm_entropy": float(np.mean(pe)),
    }


# ── small helpers (kept identical to the C# port) ────────────────────────────

def _slope(xs, ys) -> float:
    nx = len(xs)
    mx = sum(xs) / nx
    my = sum(ys) / nx
    num = 0.0
    den = 0.0
    for i in range(nx):
        dx = xs[i] - mx
        num += dx * (ys[i] - my)
        den += dx * dx
    return num / den if den != 0.0 else 0.0


def _argsort_stable(window):
    # Stable argsort by (value, index) — matches numpy argsort(kind='stable').
    return [idx for idx, _ in sorted(enumerate(window), key=lambda p: (p[1], p[0]))]


def _factorial(k: int) -> int:
    f = 1
    for i in range(2, k + 1):
        f *= i
    return f

File: eeg-backend/src/NeuroYogic.SignalProcessing/Dsp/Complexity.cs
using NeuroYogic.Domain.Analysis;

namespace NeuroYogic.SignalProcessing.Dsp;

/// <summary>
/// Non-linear complexity / entropy features — faithful C# port of the Python
/// oracle in <c>neuro_yogic/complexity.py</c>. These index the *non-oscillatory*
/// structure that distinguishes deep contemplative states (Ekagra / Niruddha),
/// which band-power and phase measures capture poorly.
///
/// Each measure is computed per channel then averaged across channels, matching
/// the band-power convention in <see cref="FeatureExtractor"/>.
/// </summary>
internal static class Complexity
{
    public static ComplexitySet Compute(double[][] filtered)
    {
        var nCh = filtered.Length;
        double lz = 0.0, hf = 0.0, se = 0.0, pe = 0.0;
        for (var c = 0; c < nCh; c++)
        {
            lz += LempelZiv(filtered[c]);
            hf += HiguchiFd(filtered[c]);
            se += SampleEntropy(filtered[c]);
            pe += PermEntropy(filtered[c]);
        }
        return new ComplexitySet(lz / nCh, hf / nCh, se / nCh, pe / nCh);
    }

    /// <summary>Normalised Lempel-Ziv complexity (LZ76) of the median-binarised signal.</summary>
    public static double LempelZiv(double[] x)
    {
        var n = x.Length;
        if (n < 2) return 0.0;
        var med = Median(x);
        var s = new int[n];
        for (var t = 0; t < n; t++) s[t] = x[t] > med ? 1 : 0;

        // LZ76 substring-counting (Kaspar & Schuster) — identical to complexity.py.
        int i = 0, c = 1, l = 1, k = 1, kMax = 1;
        while (true)
        {
            if (s[i + k - 1] == s[l + k - 1])
            {
                k++;
                if (l + k > n) { c++; break; }
            }
            else
            {
                if (k > kMax) kMax = k;
                i++;
                if (i == l)
                {
                    c++;
                    l += kMax;
                    if (l + 1 > n) break;
                    i = 0; k = 1; kMax = 1;
                }
                else
                {
                    k = 1;
                }
            }
        }
        return c * Math.Log2(n) / n;
    }

    /// <summary>Higuchi fractal dimension via the log-log slope of curve length L(k).</summary>
    public static double HiguchiFd(double[] x, int kmax = 10)
    {
        var n = x.Length;
        if (n < 4) return 0.0;
        kmax = Math.Min(kmax, n / 2);
        var lnk = new List<double>();
        var lnl = new List<double>();
        for (var k = 1; k <= kmax; k++)
        {
            var lm = new List<double>();
            for (var m = 0; m < k; m++)
            {
                var nm = (n - m - 1) / k;
                if (nm <= 0) continue;
                var acc = 0.0;
                for (var i = 1; i <= nm; i++)
                    acc += Math.Abs(x[m + i * k] - x[m + (i - 1) * k]);
                var norm = (double)(n - 1) / (nm * k);
                lm.Add(acc * norm / k);
            }
            if (lm.Count > 0)
            {
                lnl.Add(Math.Log(Mean(lm)));
                lnk.Add(Math.Log(1.0 / k));
            }
        }
        if (lnk.Count < 2) return 0.0;
        return Slope(lnk, lnl);
    }

    /// <summary>Sample entropy SampEn(m, r) with r = rCoef · std(x) (population std).</summary>
    public static double SampleEntropy(double[] x, int m = 2, double rCoef = 0.2)
    {
        var n = x.Length;
        if (n < m + 2) return 0.0;
        var r = rCoef * Std(x);
        if (r <= 0.0) return 0.0;

        int Count(int mm)
        {
            var countN = n - m;   // N - m templates for both m and m+1 (aligned)
            var c = 0;
            for (var i = 0; i < countN; i++)
                for (var j = i + 1; j < countN; j++)
                {
                    var d = 0.0;
                    for (var t = 0; t < mm; t++)
                    {
                        var diff = Math.Abs(x[i + t] - x[j + t]);
                        if (diff > d) d = diff;
                    }
                    if (d <= r) c++;
                }
            return c;
        }

        var b = Count(m);
        var a = Count(m + 1);
        if (a == 0 || b == 0) return 0.0;
        return -Math.Log((double)a / b);
    }

    /// <summary>Normalised permutation entropy of ordinal patterns (Bandt-Pompe).</summary>
    public static double PermEntropy(double[] x, int order = 3, int delay = 1)
    {
        var n = x.Length;
        var span = delay * (order - 1);
        if (n - span <= 1) return 0.0;
        var counts = new Dictionary<string, int>();
        var window = new double[order];
        for (var i = 0; i < n - span; i++)
        {
            for (var t = 0; t < order; t++) window[t] = x[i + t * delay];
            var key = string.Join(",", ArgsortStable(window));
            counts.TryGetValue(key, out var cur);
            counts[key] = cur + 1;
        }
        var total = 0;
        foreach (var v in counts.Values) total += v;
        var pe = 0.0;
        foreach (var v in counts.Values)
        {
            var p = (double)v / total;
            pe -= p * Math.Log2(p);
        }
        var denom = Math.Log2(Factorial(order));
        return denom > 0 ? pe / denom : 0.0;
    }

    // ── small helpers (kept identical to complexity.py) ─────────────────────

    private static double Median(double[] x)
    {
        var s = (double[])x.Clone();
        Array.Sort(s);
        var n = s.Length;
        return n % 2 == 1 ? s[n / 2] : (s[n / 2 - 1] + s[n / 2]) / 2.0;
    }

    private static double Std(double[] x)
    {
        var n = x.Length;
        var mean = 0.0;
        foreach (var v in x) mean += v;
        mean /= n;
        var acc = 0.0;
        foreach (var v in x) { var d = v - mean; acc += d * d; }
        return Math.Sqrt(acc / n);
    }

    private static double Mean(List<double> xs)
    {
        var s = 0.0;
        foreach (var v in xs) s += v;
        return s / xs.Count;
    }

    private static double Slope(List<double> xs, List<double> ys)
    {
        var nx = xs.Count;
        double mx = 0.0, my = 0.0;
        for (var i = 0; i < nx; i++) { mx += xs[i]; my += ys[i]; }
        mx /= nx; my /= nx;
        double num = 0.0, den = 0.0;
        for (var i = 0; i < nx; i++)
        {
            var dx = xs[i] - mx;
            num += dx * (ys[i] - my);
            den += dx * dx;
        }
        return den != 0.0 ? num / den : 0.0;
    }

    private static int[] ArgsortStable(double[] w)
    {
        var idx = new int[w.Length];
        for (var i = 0; i < w.Length; i++) idx[i] = i;
        Array.Sort(idx, (a, b) =>
        {
            var cmp = w[a].CompareTo(w[b]);
            return cmp != 0 ? cmp : a.CompareTo(b);
        });
        return idx;
    }

    private static long Factorial(int k)
    {
        long f = 1;
        for (var i = 2; i <= k; i++) f *= i;
        return f;
    }
}
